### PR TITLE
fix clusterRatelimit with more than one ratelimit instance

### LINKED ISSUE #903: Multiple Rate Limits for Same Backend are clashing
For the scenario where you try to apply two different rate limits to two separate routes for a single backend. The rate limits appear to be clashing.

Two skipper routes are created:
Route A:
Path("/routeA") && Method("GET") -> clusterClientRatelimit(1800, "1m")
Route B:
Path("/routeB") && Method("GET") -> clusterClientRatelimit(450, "1m")

**Load Test when only one rate limit is present. (~30 req/s going to /routeA)**
screen shot 2019-01-10 at 11 41 45

**Load Test when both rate limits are present. (~30 req/s evenly spread between /routeA and /routeB)**
screen shot 2019-01-10 at 11 41 52

### PR DESCRIPTION
**Pull request by szuecs**
fix #903 clusterRatelimit with more than one ratelimit instance, which overwrote the shared data from other ratelimit instances.

Signed-off-by: Sandor Szücs <sandor.szuecs@zalando.de>

### CODE REVIEW

File: ratelimit/ratelimit.go [hunk 1]
```diff
@@ -299,8 +300,9 @@ func newRatelimit(s Settings, sw Swarmer) *Ratelimit {
 		s.CleanInterval = 0
 		fallthrough
 	case ClusterClientRatelimit:
+		f := fnv.New32()
 		if sw != nil {
-			impl = newClusterRateLimiter(s, sw, "CRL")
+			impl = newClusterRateLimiter(s, sw, string(f.Sum([]byte(s.String()))))
```
Comment (aryszka) [inline ratelimit/ratelimit.go]: looks fine, just why the hashing, why not simply s.String()?

**Comment by aryszka** [comment]
:+1:

**Comment by aryszka** [comment]
as we discussed the PR in its current form doesn't fully fix the problem

**Comment by aryszka** [comment]
:-1:

**Comment by szuecs** [comment]
After an internal discussion we agreed on having the implementation changed to https://github.com/zalando/skipper/issues/912#issuecomment-454781662 which enables full control of grouping ratelimiters across routes.
The user interface would get another option to set the `group`.

**Comment by szuecs** [comment]
@aryszka the code was changed as discussed with internal customers and based on your comment https://github.com/zalando/skipper/issues/912#issuecomment-454781662.

If the code is good in the review and tested successfully in a dev cluster, I would fix all docs according to the implementation.

File: proxy/proxy.go [hunk 2]
```diff
@@ -815,30 +815,37 @@ func (p *Proxy) checkRatelimit(ctx *context) (ratelimit.Settings, int) {
...
-	}
+		s := setting.Lookuper.Lookup(ctx.Request())
+		if s == "" {
+			p.log.Errorf("Lookuper found no data in request for setting: %s and request: %v", setting, ctx.Request())
+			continue
+		}
 
-	if rl.Allow(s) {
-		return settings, 0
+		allow = allow && rl.Allow(s)
```
Comment (aryszka) [inline proxy/proxy.go]: i think the `allow` var is not required. It could be just:

```
if !rl.Allow(s) {
    return setting, rl.RetryAfter(s)
}
```

**Comment by aryszka** [comment]
lgtm

please update the branch once #932 gets merged

File: proxy/proxy.go (hunk 2 above)
Comment (szuecs) [reply proxy/proxy.go]: fixed

**Comment by szuecs** [comment]
:+1:

**Comment by aryszka** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

**Pull request merged**